src/test/shell/bazel/path_mapping_test.sh

```diff
@@ -325,4 +325,223 @@ EOF
+}
+EOF
+
+  mkdir -p "$pkg"/common/utils
+  cat > "$pkg/common/utils/BUILD" <<'EOF'
+load(":defs.bzl", "greeting_setting")
+
+greeting_setting(
+    name = "greeting",
+    build_setting_default = "Hello",
+)
+genrule(
+    name = "gen_header",
+    srcs = ["utils.h.tpl"],
+    outs = ["dir/utils.h"],
+    cmd = "cp $< $@",
+)
+genrule(
+    name = "gen_source",
+    srcs = ["utils.cc.tpl"],
+    outs = ["dir/utils.cc"],
+    cmd = "sed -e 's/{GREETING}/$(GREETING)/' $< > $@",
+    toolchains = [":greeting"],
+)
+cc_library(
+    name = "utils",
+    srcs = ["dir/utils.cc"],
+    hdrs = ["dir/utils.h"],
+    include_prefix = "other_dir",
+    strip_include_prefix = "dir",
+    visibility = ["//visibility:public"],
+)
+EOF
+  cat > "$pkg/common/utils/utils.h.tpl" <<'EOF'
+#ifndef SOME_PKG_UTILS_H_
+#define SOME_PKG_UTILS_H_
+
+#include <string>
+
+std::string AsGreeting(const std::string& name);
+#endif
+EOF
+  cat > "$pkg/common/utils/defs.bzl" <<EOF
+def _greeting_setting_impl(ctx):
+    return platform_common.TemplateVariableInfo({
+        "GREETING": ctx.build_setting_value,
+    })
+
+greeting_setting = rule(
+    implementation = _greeting_setting_impl,
+    build_setting = config.string(),
+)
+
+def _greeting_transition_impl(settings, attr):
+    return {"//$pkg/common/utils:greeting": attr.greeting}
+
+greeting_transition = transition(
+    implementation = _greeting_transition_impl,
+    inputs = [],
+    outputs = ["//$pkg/common/utils:greeting"],
+)
+
+def _transition_wrapper_impl(ctx):
+    out = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.symlink(output = out, target_file = ctx.executable.target, is_executable = True)
+    return [
+        DefaultInfo(executable = out),
+    ]
+
+transition_wrapper = rule(
+    cfg = greeting_transition,
+    implementation = _transition_wrapper_impl,
+    attrs = {
+        "greeting": attr.string(),
+        "target": attr.label(
+            cfg = "target",
+            executable = True,
+        ),
+    },
+    executable = True,
+)
+EOF
+  cat > "$pkg/common/utils/utils.cc.tpl" <<'EOF'
+#include "utils.h"
+
+std::string AsGreeting(const std::string& name) {
+  return "{GREETING}, " + name + "!";
+}
+EOF
+
+  bazel run \
+    --verbose_failures \
+    --experimental_output_paths=strip \
+    --modify_execution_info=CppCompile=+supports-path-mapping \
+    --remote_executor=grpc://localhost:${worker_port} \
+    "//$pkg:main" &>"$TEST_log" || fail "Expected success"
+
+  expect_log 'Hello, lib1!'
+  expect_log 'Hello, lib2!'
+  expect_not_log 'remote cache hit'
+
+  bazel run \
+    --verbose_failures \
+    --experimental_output_paths=strip \
+    --modify_execution_info=CppCompile=+supports-path-mapping \
+    --remote_executor=grpc://localhost:${worker_port} \
+    "//$pkg:transitioned_main" &>"$TEST_log" || fail "Expected success"
+
+  expect_log 'Hi there, lib1!'
+  expect_log 'Hi there, lib2!'
+  # Compilation actions for lib1, lib2 and main should result in cache hits due
```

Comment:
@comius This test should serve as a representative example of what path mapping can do, but please let me know if there is any additional context I can provide to help you with the review.